### ISSUE TITLE
Builtin counters of instances of cell types

### ISSUE THREAD

**Issue by omarzd** [kframework/k-legacy#1836]
Hello,

Can we implement a builtin counter for threads? that is to provide a count of the number of threads in the environment. In general, we should be able to count instances of any type of cell. This will give more power to the conditions of a rule. For instance, this will allow us to say "If ALL threads in the environment are so and so..."

I suggested a builtin counter before for transitions here (#1385).

**Comment by dwightguth** [comment]
This should be doable I think... You can use the `#configuration` function to get the complete configuration from inside a function, and then you can traverse the configuration until you get to the cell you care about and iterate through all the elements in the collection summing up the count. I suspect we will not provide builtin support for this using a hook, however.

**Comment by omarzd** [comment]
I've looked for a use of `#configuration` to know how to do that, but I didn't find any, let alone how to traverse it!

**Comment by omarzd** [comment]
Right now, I'm implementing counters crudely. All this is just a big work-around only because I need a condition that says "if no thread contains such and such". And that itself is needed to enforce precedence among rules #469.

**Comment by bmmoore** [comment]
If that's what you want, you should be able to do it with `#configuration` without needed to count anything. Maybe something like

```
anyBadThread(<top>...<thread>MyPattern</thread>...</top>) => true
anyBadThread(_) => false [owise]
```

then a side condition like `notBool(anyBadThread(#configuration()))`

**Comment by omarz1** [comment]
OK. I'm using version 3.5 (built dec 20, 2014) and it reports a syntax error whens seeing the `#` of `#configuration`.
I don't want to update and lose Maude LTLMC because I'm using that. What should I do?

**Comment by omarzd** [comment]
I downloaded version 3.6 to use `#configuration`.

In the side condition it should be `#configuration` without parentheses. 

I did the syntax definition as `syntax Bool ::= anySiteCall(Bag) [function]`.

 I also discovered that in defining the function, I must start from `<generatedTop>` like so:

```
rule anyBadThread(<generatedTop><T>... <threads>... 
             <thread>... pattern ...</thread> 
             ...</threads> ...</T></generatedTop>) => true
```

**Comment by omarz1** [comment]
One more question: the pattern inside `anyBadThread(pattern)`, must it be from the `top` cell? Can't I just do `anyBadThread(<thread>... pattern ...</thread>)`?

Does this `owise` attribute really work?

**Comment by omarz1** [comment]
This rule seems to always return false. I must be doing something wrong!

**Comment by omarz1** [comment]
I removed the [owise] false rule and I got `true` where I expected. But this means it will get stuck when it should return false. I also discovered that I must start from `<generatedTop>` or else I get a `StackOverflowError`.

**Comment by omarzd** [comment]
`[owise]` doesn't appear to be processed. It's as if the `rule anyBadThread(_) => false [owise]` is just `anyBadThread(_) => false` because it's always returning false! The `true` part is correct. I tested it. What am I doing wrong?

**Comment by omarzd** [comment]
I need this, please.

**Comment by cos** [comment]
@omarz1, could you please add a separate issue for `[owise]` and assign it to me? I don't know if someone can make it (or if it should already) work in K 3.6, but I want to make sure we'll handle it in 4.0.

**Comment by omarzd** [comment]
@cos, Done.

**Comment by omarzd** [comment]
Closed because resolved in #1870. Thank you very much @dwightguth, @bmmoore, and @cos.

**Comment by omarzd** [comment]
I want to add that because `#configuration` was causing `--search` to crash (as reported in #1874), I got rid of it and instead of passing my functions `#configuration` as an argument, I pass it the bag of threads referenced in the rule like so:

```
rule 
  <threads>
    A:Bag
  </threads>
some stuff => some other stuff
when notBool anyBadThread(A)
```

And in case the rewrite is  happening inside a thread, then:

```
rule
  <threads>
    A:Bag
    <thread>
      some stuff => some other stuff  // also ensure that this thread is not a bad thread as well
    </thread>
  </threads>
when notBool anyBadThread(A)
```
